### ISSUE TITLE
.NET Core 2.1 Roslyn cannot find System.Decimal in assemblies

### ISSUE THREAD

**Issue by justinushermawan** [dotnet/core#2082]
I'm using **Roslyn** to compile my class in runtime to `assembly` output. I've added my needed assemblies into metadata references as below:

```c#
string p = Path.GetDirectoryName(typeof(object).Assembly.Location);
_ref = new MetadataReference[] {
     MetadataReference.CreateFromFile(Path.Combine(p, "System.Collections.dll")),
     MetadataReference.CreateFromFile(Path.Combine(p, "System.Runtime.dll")),
     MetadataReference.CreateFromFile(Path.Combine(p, "System.Private.CoreLib.dll")),
     MetadataReference.CreateFromFile(typeof(object).Assembly.Location),          
     MetadataReference.CreateFromFile(typeof(Schema).Assembly.Location)
};
```
And my CSharp syntax tree:

```c#
using System;

namespace _
{
    public sealed class Model : Record
    {
        public string name;
        public decimal amount;
        public bool granted;
        public DateTime birthdate;
    }
}
```
And when I use my `dll` file in **another project**, I got an error when I create an instance of `Model`, the error is located on my `decimal` property which is `amount`. Omnisharp shows me:

`The type 'Decimal' is defined in an assembly that is not referenced. You must add a reference to assembly 'System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e'.`

As far as I known, the `System.Private.CoreLib.dll` has been existed in `typeof(object).Assembly.Location`, so it doesn't make any sense if still giving me the error.

What's wrong here??

**Comment by karelz** [comment]
@jaredpar can you please help me route this issue? It looks like Roslyn specific.

**Comment by justinushermawan** [comment]
If I don't reference that DLL, I'd get an error that ask me to reference it because System.Decimal could not be found. What should I do?

**Comment by jaredpar** [comment]
What is the exact error and the full list of DLLs that are referenced? Should also include where you are referencing them from. 

It's much trickier to do this on CoreCLR than Desktop.  Have to be careful to reference the correct reference DLLs. Referencing the runtime ones, or mixing the two, will lead to a lot of issues.

**Comment by justinushermawan** [comment]
I simplified the case. I have 2 netcore projects, The first one is where Roslyn generates my Model class (contains 'string', 'int', and 'decimal' properties) into a DLL file by using FileStream approach (not MemoryStream). The metadatareference[] assemblies linked are:
'System.Runtime
System.Private.CoreLib
and typeof(MyCurrentClass).Assembly.Location'
If I run this one, it works and produces the dll of MyCurrentClass completely. 

Now I try to reference my those DLLs (MyCurrentClass.dll and MyFirstProject.dll) into the second project, I use itemgroup reference in my .csproj file. And I try to create a simple instance of MyCurrentClass, but I got an error when I try to access the 'decimal' property of it, the other properties work fine, only the 'decimal' one.

**Comment by jaredpar** [comment]
> Now I try to reference my those DLLs (MyCurrentClass.dll and MyFirstProject.dll) into the second project,

Mentioned this in my first response though: don't reference System.Private.Corelib. It's not supported and if it works it's an accidental success.

**Comment by justinushermawan** [comment]
But, it asks me to reference it by giving me an error if I don't reference it. I even add that reference because of the error.

**Comment by jaredpar** [comment]
It asks you to give that reference because you're providing other runtime libraries instead of reference libraries. As I mentioned in my earlier comment:

> What is the exact error and the full list of DLLs that are referenced? Should also include where you are referencing them from.

The latter part will likely point to the runtime directory which is the source of the problem.

**Comment by justinushermawan** [comment]
I put the latter part because I need to inherit a class from MyProject.dll, how can I achieve that?

**Comment by karelz** [comment]
@justinushermawan run HelloWorld compilation in VS. Enable verbose msbuild logs, check the args it passes to Roslyn - you will see it uses reference assemblies, none of which is called System.Private.Corelib.
Reference assemblies is where types live at compile time. At runtime many of those type-forward to implementation assemblies (like System.Private.Corelib).
You should never see an app or library to reference the implementation assembly - if they do, it is a bug they need to fix.

**Comment by justinushermawan** [comment]
@karelz I've removed those implementation assemblies in my metadatareference, now I got the following errors on runtime:

```
Unhandled Exception: System.Exception: CS8021: No value for RuntimeMetadataVersion found. No assembly containing System.Object was found nor was a value for RuntimeMetadataVersion specified through options.
CS0246: The type or namespace name 'Record' could not be found (are you missing a using directive or an assembly reference?)
CS0518: Predefined type 'System.String' is not defined or imported
CS0518: Predefined type 'System.Int64' is not defined or imported
CS0518: Predefined type 'System.Void' is not defined or imported
CS0518: Predefined type 'System.Object' is not defined or imported
CS0518: Predefined type 'System.Int64' is not defined or imported
CS0518: Predefined type 'System.Object' is not defined or imported
CS0246: The type or namespace name 'Record' could not be found (are you missing a using directive or an assembly reference?)
CS0518: Predefined type 'System.Int64' is not defined or imported
CS0518: Predefined type 'System.String' is not defined or imported
CS0518: Predefined type 'System.Object' is not defined or imported
CS0246: The type or namespace name 'Record' could not be found (are you missing a using directive or an assembly reference?)
CS0115: 'Employee.Write(string, long)': no suitable method found to override
CS0115: 'Employee.Read(SortedList<long, Record>, out long, string)': no suitable method found to override
CS0518: Predefined type 'System.String' is not defined or imported
CS0518: Predefined type 'System.String' is not defined or imported
CS0518: Predefined type 'System.Decimal' is not defined or imported
CS0518: Predefined type 'System.Int32' is not defined or imported
CS0518: Predefined type 'System.Object' is not defined or imported
CS0246: The type or namespace name 'DateTime' could not be found (are you missing a using directive or an assembly reference?)
CS0518: Predefined type 'System.Object' is not defined or imported
CS0246: The type or namespace name 'BinaryWriter' could not be found (are you missing a using directive or an assembly reference?)
CS0518: Predefined type 'System.Object' is not defined or imported
CS0246: The type or namespace name 'BinaryWriter' could not be found (are you missing a using directive or an assembly reference?)
CS0518: Predefined type 'System.Object' is not defined or imported
CS0246: The type or namespace name 'FileStream' could not be found (are you missing a using directive or an assembly reference?)
CS0518: Predefined type 'System.String' is not defined or imported
CS0518: Predefined type 'System.Object' is not defined or imported
CS0518: Predefined type 'System.Int32' is not defined or imported
CS0518: Predefined type 'System.Object' is not defined or imported
CS0103: The name 'DateTime' does not exist in the current context
CS0518: Predefined type 'System.String' is not defined or imported
CS0518: Predefined type 'System.Object' is not defined or imported
CS0103: The name 'FileMode' does not exist in the current context
CS0518: Predefined type 'System.Object' is not defined or imported
CS0518: Predefined type 'System.Object' is not defined or imported
CS1061: 'Employee' does not contain a definition for 'next' and no accessible extension method 'next' accepting a first argument of type 'Employee' could be found (are you missing a using directive or an assembly reference?)
CS0518: Predefined type 'System.Object' is not defined or imported
CS0518: Predefined type 'System.Boolean' is not defined or imported
CS0518: Predefined type 'System.Boolean' is not defined or imported
CS0518: Predefined type 'System.Object' is not defined or imported
CS1061: 'Employee' does not contain a definition for 'id' and no accessible extension method 'id' accepting a first argument of type 'Employee' could be found (are you missing a using directive or an assembly reference?)
CS0518: Predefined type 'System.Object' is not defined or imported
CS1061: 'Employee' does not contain a definition for 'state' and no accessible extension method 'state' accepting a first argument of type 'Employee' could be found (are you missing a using directive or an assembly reference?)
CS0518: Predefined type 'System.Object' is not defined or imported
CS1061: 'Employee' does not contain a definition for 'timestamp' and no accessible extension method 'timestamp' accepting a first argument of type 'Employee' could be found (are you missing a using directive or an assembly reference?)
CS0518: Predefined type 'System.Object' is not defined or imported
CS1061: 'Employee' does not contain a definition for 'memento' and no accessible extension method 'memento' accepting a first argument of type 'Employee' could be found (are you missing a using directive or an assembly reference?)
CS0518: Predefined type 'System.Object' is not defined or imported
CS0518: Predefined type 'System.Object' is not defined or imported
CS0246: The type or namespace name 'BinaryReader' could not be found (are you missing a using directive or an assembly reference?)
CS0518: Predefined type 'System.Object' is not defined or imported
CS0246: The type or namespace name 'BinaryReader' could not be found (are you missing a using directive or an assembly reference?)
CS0518: Predefined type 'System.Object' is not defined or imported
CS0246: The type or namespace name 'FileStream' could not be found (are you missing a using directive or an assembly reference?)
CS0518: Predefined type 'System.Object' is not defined or imported
CS0103: The name 'FileMode' does not exist in the current context
CS0518: Predefined type 'System.Int64' is not defined or imported
CS0518: Predefined type 'System.Int32' is not defined or imported
CS0518: Predefined type 'System.Boolean' is not defined or imported
CS0518: Predefined type 'System.Boolean' is not defined or imported
CS0518: Predefined type 'System.Object' is not defined or imported
CS0518: Predefined type 'System.Object' is not defined or imported
CS0103: The name 'Activator' does not exist in the current context
CS0518: Predefined type 'System.Object' is not defined or imported
CS0518: Predefined type 'System.Type' is not defined or imported
CS0518: Predefined type 'System.Boolean' is not defined or imported
CS0518: Predefined type 'System.Object' is not defined or imported
CS0518: Predefined type 'System.Object' is not defined or imported
CS1061: 'Employee' does not contain a definition for 'next' and no accessible extension method 'next' accepting a first argument of type 'Employee' could be found (are you missing a using directive or an assembly reference?)
CS0518: Predefined type 'System.Object' is not defined or imported
CS1061: 'Employee' does not contain a definition for 'id' and no accessible extension method 'id' accepting a first argument of type 'Employee' could be found (are you missing a using directive or an assembly reference?)
CS0518: Predefined type 'System.Object' is not defined or imported
CS1061: 'Employee' does not contain a definition for 'state' and no accessible extension method 'state' accepting a first argument of type 'Employee' could be found (are you missing a using directive or an assembly reference?)
CS0518: Predefined type 'System.Object' is not defined or imported
CS1061: 'Employee' does not contain a definition for 'timestamp' and no accessible extension method 'timestamp' accepting a first argument of type 'Employee' could be found (are you missing a using directive or an assembly reference?)
CS0518: Predefined type 'System.Object' is not defined or imported
CS1061: 'Employee' does not contain a definition for 'memento' and no accessible extension method 'memento' accepting a first argument of type 'Employee' could be found (are you missing a using directive or an assembly reference?)
CS0518: Predefined type 'System.Object' is not defined or imported
CS0246: The type or namespace name 'DateTime' could not be found (are you missing a using directive or an assembly reference?)
CS0518: Predefined type 'System.Object' is not defined or imported
CS1061: 'Employee' does not contain a definition for 'memento' and no accessible extension method 'memento' accepting a first argument of type 'Employee' could be found (are you missing a using directive or an assembly reference?)
CS0518: Predefined type 'System.Object' is not defined or imported
CS0103: The name 'Memento' does not exist in the current context
CS0518: Predefined type 'System.Boolean' is not defined or imported
CS0518: Predefined type 'System.Boolean' is not defined or imported
CS0518: Predefined type 'System.Object' is not defined or imported
```

Note that the `Record` class is located in my parent project.
How to handle it?

**Comment by karelz** [comment]
Is that HelloWorld app? Did you try to mimic what VS does with the references?

**Comment by justinushermawan** [comment]
Yes, I tried with 2 simple projects, the one that produced the DLL of simple class, and the second one that using the DLL, it still doesn't work.

**Comment by karelz** [comment]
ok, can you share full repro for the HelloWorld-style repro?

**Comment by justinushermawan** [comment]
Sure, https://github.com/justinushermawan/roslyn-test

**Comment by karelz** [comment]
This is wrong:
https://github.com/justinushermawan/roslyn-test/blob/master/Core/Program.cs#L28

You're using implementation DLL (the one you're running on), not reference assemblies as you should.

**Comment by justinushermawan** [comment]
What should I change it with? I tried netstandard.dll, still not working.

**Comment by justinushermawan** [comment]
I've compiled in VS, where can I find the reference assemblies directory?
SS:
![1](https://user-images.githubusercontent.com/13159271/49197761-b921b880-f3c2-11e8-8806-057d0d10ec25.png)

**Comment by justinushermawan** [comment]
These one?
![2](https://user-images.githubusercontent.com/13159271/49197965-9e9c0f00-f3c3-11e8-919a-4665ec1281b8.png)

**Comment by justinushermawan** [comment]
It works when I use:
```
string p = @"C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.netcore.app\2.1.0\ref\netcoreapp2.1\";
MetadataReference[] _ref = new MetadataReference[] {
        MetadataReference.CreateFromFile(Path.Combine(p, "netstandard.dll")),
        MetadataReference.CreateFromFile(Path.Combine(p, "System.dll")),
        MetadataReference.CreateFromFile(Path.Combine(p, "System.Runtime.dll"))
};
```

Is it the right way?

**Comment by dasMulli** [comment]
I believe the "absolutely right" way would be to make use of compilation context preservation.

1. In your csproj, set
```xml
  <PropertyGroup>
    <PreserveCompilationContext>true</PreserveCompilationContext>
  </PropertyGroup>
```
2. Add package reference to `Microsoft.Extensions.DependencyModel`:
```xml
  <ItemGroup>
    <PackageReference Include="Microsoft.Extensions.DependencyModel" Version="2.1.0" />
  </ItemGroup>
```
3. Use `DependencyContext` to get compilation references:
```csharp
using Microsoft.Extensions.DependencyModel;
using System.Linq;
…
MetadataReference[] _ref = 
    DependencyContext.Default.CompileLibraries
    .SelectMany(cl => cl.ResolveReferencePaths())
    .Select(asm => MetadataReference.CreateFromFile(asm))
    .ToArray();
```

**Comment by justinushermawan** [comment]
@dasMulli Yeah, it works. Thank you.

Would you please explain my code mistakes?

**Comment by dasMulli** [comment]
> Would you please explain my code mistakes?

Not a mistake, there's just an awful lot of things that can go wrong and you'd need detailed information about how either compilation or the layering of runtime DLLs work.

`PreserveCompilationContext` makes sure that your `.deps.json` contains the references that your app has during compilation. When you build your app, the compile-time assemblies come from the `Microsoft.NETCore.App` NuGet package. Specifically, from
```
.nuget/packages/microsoft.netcore.app/2.1.0/ref/netcoreapp2.1/*.dll
```

There also is a set of MSBuild targets that resolves conflicts of NuGet packages that contain implementations for older versions of .NET Core that are now part of .NET Core (specifically, the `System.*` NuGet packages from the 1.* timeframe).
The preserved compilation context will also contain the filtered results so you're not going to run into conflicts.

`DependencyContext` will help you read the .deps.json at runtime and resolve the paths to the dll files you need.

**Comment by justinushermawan** [comment]
Does `PreserveCompilationContext` make my build time slower? And if I don't set it to `true`, can't DependencyContext work?

**Comment by dasMulli** [comment]
> Does PreserveCompilationContext make my build time slower? 

May be a bit, but it shouldn't matter.

> And if I don't set it to true, can't DependencyContext work?

Nope, it will throw exceptions. And you'll be missing the dlls once published.

**Comment by justinushermawan** [comment]
Oh I see. Thanks in advance for your help.

**Comment by codeex** [comment]
Thanks ！ @dasMulli

**Comment by ddobrev** [comment]
@dasMulli are you absolutely sure this is the only way to properly compile C# with Roslyn? I can't believe these reference assemblies can only be accessed with a project setting and an external NuGet dependency. This means Roslyn is completely incapable of compiling even the simplest program on its own. What good is it then?

**Comment by Lunar2kPS** [comment]
Strange, @dasMulli 's solution doesn't work anymore (at least for me)... maybe cause I'm on .NET 6 now.

I'm lost trying to get Roslyn to compile an executable that actually runs without all these assembly referencing errors... it has so many different ways of saying how System.Private.CoreLib.dll isn't found, or Object is undefined somehow, it's just strange..

There has to be a way for Roslyn to compile C# code for .NET 5+, otherwise it wouldn't be so widely used nowadays, right?
Am I missing something?

**Comment by MSDN-WhiteKnight** [comment]
>Strange, dasMulli 's solution doesn't work anymore (at least for me)... maybe cause I'm on .NET 6 now.

In which way it does not work? Throws exception, or produces incorrect binaries?

>There has to be a way for Roslyn to compile C# code for .NET 5+, otherwise it wouldn't be so widely used nowadays, right?
Am I missing something?

Commenters on this thread might be using confusing terms. Roslyn is widely used in general, but using Roslyn Emit API directly is not.

**Comment by Lunar2kPS** [comment]
> In which way it does not work? Throws exception, or produces incorrect binaries?

So if I followed steps 1-3 exactly (not the .NET Core-only addition after step 3),
I get the following:

```
Unhandled Exception: System.IO.FileNotFoundException: Could not load file or assembly 'System.Runtime, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
````

If I also add this code, where `refs` is my `List<MetadataReference>`,
```cs
Assembly.GetEntryAssembly().GetReferencedAssemblies()
    .ToList()
    .ForEach(a => {
        Console.WriteLine(Assembly.Load(a).Location);
        refs.Add(MetadataReference.CreateFromFile(Assembly.Load(a).Location));
    });
```

I still get the same error.

✅ To be clear, the compilation is successful,
❌ But I get these errors when running the resultant executable (by double-clicking on the .exe file on Windows, or by using Process.Start(...) on it).

> 
> Commenters on this thread might be using confusing terms. Roslyn is widely used in general, but using Roslyn Emit API directly is not.

Ohh wow.. why is it different, doesn't the C# Roslyn compilation/emit stuff _(that I'm trying to do, but getting errors when running the executable)_ use the Roslyn compiler, so shouldn't it be the same compilation workflow under the hood?

**Comment by MSDN-WhiteKnight** [comment]
Your errors don't look related to this issue. This issue is about error "The type 'Decimal' is defined in an assembly that is not referenced.". In you case, the correct assembly is referenced, but it's not found at runtime. Maybe something related to self-contained/framework-dependent app differences. It's best to open new issue, and include more code, showing how you invoke Emit API with these references. This issue is closed, so you are unlikely to get much help here.

>Ohh wow.. why is it different, doesn't the C# Roslyn compilation/emit stuff (that I'm trying to do, but getting errors when running the executable) use the Roslyn compiler, so shouldn't it be the same compilation workflow under the hood?

Indeed it's the same. But when Emit API is invoked in MSBuild context, this problem does not exist, because correct reference assembly paths are always known, they are passed from the SDK. If you want to invoke Emit API in arbitrary app that can be run even on a machine without SDK at all, you need to get them somehow. The method with PreserveCompilationContext and DependencyModel solves this. But you may hit a different issue then, it's not unlikely, because again, using Emit API directly is advanced scenario.

**Comment by Lunar2kPS** [comment]
Gotcha, thanks okay! I'll open a new issue, my bad and thanks for the explanations :)
